pkg/eval/builtin_fn_io.go

```diff
@@ -194,6 +195,180 @@ func print(fm *Frame, opts printOpts, args ...interface{}) {
+// behave differently based on the associated value type. In Elvish the
+// type is always an integer; specifically, a Go `int`; regardless of whether
+// the Elvish value is a string or float64. That is, the number is coerced to
+// an int if possible else an exception is raised.
+//
+// **Note**: Do not use the `%p` or `%q` formatting verbs at this time. The `%p`
+// verb will never have any meaning. The `%q` formatting verb will eventually
+// behave as if you had printed the result of `repr` on the value. Similarly,
+// only scalars such as strings and numbers are currently supported. Passing
+// an Elvish list, map, exception, or other complex type will produce
+// unexpected output. This will be improved in the future.
+//
+// **Note**: This is loosely based on the [POSIX `printf`
+// command](https://pubs.opengroup.org/onlinepubs/007908799/xcu/printf.html).
+// There are three notable differences. The first is that the formatting verbs
+// and behavior use Go's [`fmt`](https://golang.org/pkg/fmt/) package and not
+// the POSIX specification. The two have many similarities but are not
+// identical. The second is that the number of values must match the number of
+// formatting verbs. Excess values do not result in the format being evaluated
+// additional times until all values are consumed as happens with the POSIX
+// command of the same name. You must explicitly split the value list and
+// invoke this builtin for each block of values. The third is that the Elvish
+// `printf` does not recognize backslash sequences such as `\n`. You must use
+// [double-quoted Elvish strings](language.html#double-quoted-string) if you
+// want such backslash sequences to be recognized -- just as you would with a
+// string passed to any other Elvish builtin.
+//
+// @cf print echo pprint repr
+
+type formatterString struct {
+	str string
+}
+type formatterFloat64 struct {
+	num float64
+}
+
+func printf(fm *Frame, template string, args ...interface{}) {
+	out := fm.OutputFile()
+	aliasedArgs := make([]interface{}, len(args))
+	// TODO: Implement support for `%q` but using Elvish's rules for quoting
+	// strings such as produced by `repr '\a"b'"\033'"` rather than Go's
+	// default `%q` behavior.
+	//
+	// TODO: Possibly special case other Elvish types such as lists, maps and
+	// exceptions.
+	for i, a := range args {
+		switch a := a.(type) {
+		case string:
+			aliasedArgs[i] = formatterString{a}
+		case float64:
+			aliasedArgs[i] = formatterFloat64{a}
+		default:
+			// This will pretty much only work if the `%v` fmt verb is used or
+			// the argument type matches the formatting verg (e.g., passing
+			// `$true` to the `%t` verb).
+			aliasedArgs[i] = a
+		}
+	}
+	s := fmt.Sprintf(template, aliasedArgs...)
+	out.WriteString(s)
+}
+
+// Convert a formatting verb state to a string that is logically equivalent to
+// the original formatting verb.
+func buildFmtVerb(state fmt.State, v rune) string {
+	var flags = []rune{'%'}
```

Comment:
This function uses a variety of ways to build up the string (`append`, `fmt.Sprintf` and `+`); use a [strings.Builder](https://godoc.org/strings#Builder) instead.